### DISCUSSION TITLE
Map attribute to key (pk/sk)

### DISCUSSION THREAD

**Discussion by shishkin** [sensedeep/dynamodb-onetable#229] [Q&A] [answered]
Hello! Thanks a lot for this comprehensive library!

I'm trying to scan a GSI that I only want to project keys from. I have an attribute that is part of the index key mapping, e.g. `gs1sk: { type: String, value: "${updated}" }`. No I wonder if it's possible to retrieve the value of the projected sort key by the logical attribute name which is `updated`? It works if I either project all or specific fields, enable following or access return value by `gs1sk`, but I would prefer to avoid any of those in order to keep the index storage smaller, table index definition decoupled from logical entities and my queries operate on logical abstraction level.

UPDATED:

Here is the schema:

```
Updates: {
  gs1pk: { type: String, value: "${state}" },
  gs1sk: { type: String, hidden: true },
  state: { type: String, required: true },
  updated: { type: Date, required: true, map: "gs1sk" },
  ... 
}
```

Now I can get last updated date by state like so:

```
const result = await updatesModel.scan(
  { state: state },
  {
    index: "GS1",
    limit: 1,
    reverse: true,
  }
);
console.log("Result:", result[0].updated);
```

This works in this simple example where the `updated` field is directly mapped to a key. But what if I want to create a composite key out of multiple attributes, e.g. `gs1sk: { type: String, value: "${key1}#${key2}#${key3}" }`? It would be nice to be able to access scan results like `result[0].key1, result[0].key2, result[0].key3`. I understand this would require parsing the key to extract key parts.

**Comment by mobsense** [comment]
Can you please provide a code example of what you are trying to do -- that will clear up some ambiguities.

Guessing ... you are trying to do a get / find like:

    await db.find('Model', {foo: 'bar'}, {index: 'gs1'})

And you'd like to use GS1 which projects only keys. If the schema defines the foo property and it is used in the value template for the gs1sk, then OneTable should construct the gs1sk value to retrieve based on the value of "foo" in the API call.

**Reply by shishkin** [reply]
Sorry for not providing an example right away. I've updated the question above.

**Comment by mobsense** [comment, marked as answer]
Some comments:

* You are using scan. Note: scan does not use the keys and will scan the entire table. You probably want "find" rather than scan.
* Your example uses gs1pk: { type: String, value: "${state}" }, but your text refers to value: "${updated}". I presume you mean ${updated}?

Your return results will not reverse parse fields that use value templates with multiple components like `gs1sk: { type: String, value: "${key1}#${key2}#${key3}" }`

You return results are always just the top level fields. Value templates are input only. However you should get all the returned components key1, key2, key3 if they are projected to the index and they are defined in the schema.

**Reply by shishkin** [reply]
I'm using a scan with a hash key `state`, `limit: 1` and revers sort by the sort key `updated` in order to find the last updated date for state. Actually both fields are part of GS1 index, so no additional projection should be necessary give a reverse map from keys to logical attributes.

**Reply by mobsense** [reply]
Can you please post a complete test case with schema and actual code that fails that we can run to reproduce. That will make it much easier to answer. Please check with version 2.2.1 and post the error messages. The issue template has guidance for all the required fields.
